// scaffdog/dist/index.mjs
import { createRequire } from 'module';
import { createConsola, LogLevels } from 'consola';
import { a as createScaffdogInitializer, c as createLibrary } from './shared/scaffdog.23370b3c.mjs';
export { ScaffdogAggregateError, ScaffdogError } from '@scaffdog/error';
import 'path';
import '@scaffdog/config';
import '@scaffdog/engine';
import 'vm';
import 'is-plain-obj';
import '@scaffdog/core';
import 'front-matter';
import 'zod';
import 'chalk';
import 'indent-string';
import 'fs';
import 'fs/promises';
import 'globby';
import 'fuse.js';
import 'inquirer';
import 'inquirer-autocomplete-prompt';

const require = createRequire(import.meta.url);
const pkg = require("../package.json");
const logger = createConsola({
  level: LogLevels.silent,
  reporters: []
});
const lib = createLibrary(logger);
const initializer = createScaffdogInitializer({
  lib
});
const version = pkg.version;
const createScaffdog = initializer.createScaffdog;
const loadScaffdog = initializer.loadScaffdog;

export { createScaffdog, loadScaffdog, version };
